### ISSUE TITLE
TypeOnly optimization in RowExpression can cause wrong result in nested cast

### ISSUE THREAD

**Issue by hellium01** [prestodb/presto#13116]
Following queries:
```
select cast(varchar_value as varchar(3)) || ' sfd' from (values ('9898.122')) t(varchar_value) ;
select cast(cast(varchar_value as varchar(3)) as varchar(5)) from (values ('9898.122')) t(varchar_value);
```

will return:
```
9898. sfd
and 
9898.
```
instead of the expected:
```
989 sfd
989
```
Since the outside cast is typeOnly, the optimization inside `SqlToRowExpressionTranslator` will remove the outside cast and change the inside cast into cast as VARCHAR(5). 

This is likely a long term bug got triggered by recent code path changes.

**Comment by hellium01** [comment]
The problematic code is in `SqlToRowExpressionTranslator`, line 488:
```
            if (node.isTypeOnly()) {
                return changeType(value, getType(node));
            }
```
This code path was never called previously because `node.isTypeOnly` will always be false on worker because of the serialization/deserialization of `Expression` stripped off this information. 
However, since now the translation is moved earlier in coordinator. Thus this bug is exposed.

**Comment by rongrong** [comment]
Related to #12891 The problematic code is the same but manifested in different symptoms.

**Comment by rongrong** [comment]
The problem is the definition of `typeOnly`. From the way we are using it, `typeOnly` seems to mean "it's safe to change the type of output to input and remove cast". If this is the definition, then all `CAST` in the expression tree should be coercible. Currently we say a `CAST` is type only as long as the `CAST` itself is coercible. In the above mentioned example, `CAST(varchar(3) AS varchar)` is a type only cast, however, `CAST('9898.122' as varchar(3))` is not. If we assume the outer cast is type only even though the inner cast is not, we can optimize the outer cast to have type `varchar`, which means we effectively changed the inner cast to `CAST('9898.122' as varchar)`.

**Comment by hellium01** [comment]
This is resolved by #13127
